cranelift/codegen/src/machinst/lower.rs

```diff
@@ -457,6 +595,103 @@ impl<'func, I: VCodeInst> Lower<'func, I> {
         })
     }
 
+    /// Pre-analysis: compute `value_ir_uses`. See comment on
+    /// `ValueUseState` for a description of what this analysis
+    /// computes.
+    fn compute_use_states<'a>(f: &'a Function) -> SecondaryMap<Value, ValueUseState> {
+        // We perform the analysis without recursion, so we don't
+        // overflow the stack on long chains of ops in the input.
+        //
+        // This is sort of a hybrid of a "shallow use-count" pass and
+        // a DFS. We iterate over all instructions and mark there args
+        // as used. However when we increment a use-count to
+        // "Multiple" we push its args onto the stack and do a DFS,
+        // immediately marking the whole dependency tree as
+        // Multiple. Doing both (shallow use-counting over all insts,
+        // and deep Multiple propagation) lets us trim both
+        // traversals, stopping recursion when a node is already at
+        // the appropriate state.
+        //
+        // In particular, note that the *coarsening* into {Unused,
+        // Once, Multiple} is part of what makes this pass more
+        // efficient than a full indirect-use-counting pass.
+
+        let mut value_ir_uses: SecondaryMap<Value, ValueUseState> =
+            SecondaryMap::with_default(ValueUseState::Unused);
+
+        // Stack of iterators over Values as we do DFS to mark
+        // Multiple-state subtrees.
+        type StackVec<'a> = SmallVec<[std::slice::Iter<'a, Value>; 16]>;
+        let mut stack: StackVec = smallvec![];
+
+        // Push args for a given inst onto the DFS stack.
+        let push_args_on_stack = |stack: &mut StackVec<'a>, value| {
+            log::trace!(" -> pushing args for {} onto stack", value);
+            if let ValueDef::Result(src_inst, _) = f.dfg.value_def(value) {
+                stack.push(f.dfg.inst_args(src_inst).iter());
+            }
+        };
+
+        // Do a DFS through `value_ir_uses` to mark a subtree as
+        // Multiple.
+        let process_dfs = |value_ir_uses: &mut SecondaryMap<Value, ValueUseState>,
```

Comment:
```suggestion
        let mark_all_uses_as_multiple = |value_ir_uses: &mut SecondaryMap<Value, ValueUseState>,
```

I think this might make it more clear.